### ISSUE TITLE
How to use COM812 in linter but avoid conflict with formatter?

### ISSUE THREAD

**Issue by jagerber48** [astral-sh/ruff#9216]
I would like to achieve the following things:

- Use the ruff linter on my code
- Use the ruff formatter on my code
- Enable rule COM812 for the linter

I like the linter, I like the formatter, and I like rule COM812. However, when I have COM812 as an allowed rule I get a warning when I run the formatter that rule COM812 may cause conflicts and that it should be disabled to avoid unexpected behavior. Is there, then, a recommended way to achieve my three goals?

I see a number of issues discussing rule COM812 but I don't see any addressing this exact question.

**Comment by charliermarsh** [comment]
I don't think it's currently possible based on our definition of "incompatible". In short, we want to avoid cases in which running the formatter introduces _new_ lint rule violations. In the case of COM812, you could have a function like:

```python
def args(aaaaaaaa, bbbbbbbbb, cccccccccc, ddddddddd, eeeeeeee, ffffff, gggggggggggg, hhhh):
    pass
```

Which the formatter may reformat as:

```python
def args(
    aaaaaaaa, bbbbbbbbb, cccccccccc, ddddddddd, eeeeeeee, ffffff, gggggggggggg, hhhh
):
    pass
```

Which would violate COM812. After adding the trailing comma to fix COM812, the formatter would then reformat again as:

```python
def args(
    aaaaaaaa,
    bbbbbbbbb,
    cccccccccc,
    ddddddddd,
    eeeeeeee,
    ffffff,
    gggggggggggg,
    hhhh,
):
    pass
```

We would need a new trailing-comma setting in the formatter, like `"force"`.

\cc @MichaReiser

**Comment by jagerber48** [comment]
Thanks for the response! I would be interested in any sort of solution to get compatibility, something like the `"force"` flag you mention seems fine to me. I would be curious to know if it's decided that this is something `ruff` will support at some point.

I think I've already experienced the two step process you described I think for now, for my codebase, I will bear the annoyance of the incompatibility message and implement the `format - lint (fix) - format` routine to make the formatting stable.

**Comment by bittner** [comment]
I'm refactoring large code bases, and the case you quote above is particularly common as an outcome of reformatting.

```python
def args(
    aaaaaaaa, bbbbbbbbb, cccccccccc, ddddddddd, eeeeeeee, ffffff, gggggggggggg, hhhh
):
    ...
```

What the formatter does is indeed very natural and intuitive: When the line gets too long it first tries to put everything inside braces onto a line of its own.  Shouldn't COM812 accept this as a good case, in general? Kind of, "there are several arguments on a single line", that's a signal that it's okay not to have a trailing comma.

The `trailing-comma = "force"` is probably just that. - Maybe `"strict"` vs. `"forgiving"` would reflect the intention even better, with `"forgiving"` as the default, probably.

**Comment by MichaReiser** [comment]
I understand that having `COM812` and the formatter not agreeing on the formatting here is annoying. 

What's the motivation for using both the formatter and `COM812`. Wouldn't it be sufficient to only rely on the formatter that also fixes the issue automatically rather than having the lint rule enabled too? Or do you need `COM812` enabled because you dislike the "all on one line" formatting that the formatter produces?

**Comment by jagerber48** [comment]
> What's the motivation for using both the formatter and COM812. Wouldn't it be sufficient to only rely on the formatter that also fixes the issue automatically rather than having the lint rule enabled too? Or do you need COM812 enabled because you dislike the "all on one line" formatting that the formatter produces?

There are a few things I want to satisfy. The motivation for `COM812` from [the docs](https://docs.astral.sh/ruff/rules/missing-trailing-comma/) is
> The presence of a trailing comma can reduce diff size when parameters or elements are added or removed from function calls, function definitions, literals, etc.

This is true in cases where there is one argument per line. So in those cases (one argument per line) I do want `COM812` enforced.

Another constraint is I want arguments to be wrapped in some way when a single line like
```
def args(aaaaaaaa, bbbbbbbbb, cccccccccc, ddddddddd, eeeeeeee, ffffff, gggggggggggg, hhhh):
    pass
```
is too long.
But I would be happy with
```
def args(
    aaaaaaaa, bbbbbbbbb, cccccccccc, ddddddddd, eeeeeeee, ffffff, gggggggggggg, hhhh
):
    pass
```
As a resolution to a line too long in the case that the arguments are too long to stand on one like with the function definition (`def args( ... )`) but short enough to stand on their own line. That is, I don't think this case needs a trailing comma because I don't think it shortens the diff.

So yes, I think I agree with @bittner that I don't understand why this form violates `COM812`. Maybe `COM812` should only be enforced when there is one argument per line. That way the trailing comma won't be added by the linter after the re-formatting, thus inducing the formatter to again want to do more.

This wouldn't even require a new option I don't think. BUT if a new option is the way to go: Is it possible for LINTING rules to have optional configuration (other than inclusion or exclusion)? Or can only formatting rules be controlled by optional configuration?

**Comment by MichaReiser** [comment]
I'm currently looking into what it takes to make `COM812` compatible with the formatter. It sounds like a simple change, but it's a very expensive token-based rule. I need to figure out how to test if all arguments (or list elements etc) are on their own line without having to track too much state.

**Comment by MichaReiser** [comment]
Interesting, the existing rule isn't able to catch all cases where I would suspect a trailing comma to be required because it only tests if there's a newline between the last token and the closing parentheses:

E.g. the following is okay

```python
def call(a, b,
    c): pass
```

**Comment by jagerber48** [comment]
Interesting. As for my desired behavior independent of any existing functionality: Assuming all arguments are simple parameter names, the three types of formats I would ever want (in this preference) are

- One line, params/args appear on the same line as `def foo(...)`
- Three lines. One line for the function definition, one line for params and args and one line for the closing parentheses, colon, and return type annotation.
- n lines. One line for the function definition, one line for each param/arg pair, and one line for the closing parentheses, colon, and return type annotation.

I would prefer no trailing comma for the first two but not be upset by the existence of a trailing comma. The third one must have a trailing comma.

Anything other than these three options, like what was in the last post, would not be acceptable to me.

I don't have a simple way to say what rules I expect if there is nesting of complex expressions
```
def foo(
    param_1=3
    param_2=[
        first_element,
        bar(
            "a deeply nested input",        
        ),
    ],
    param_3="top_layer_input",                  
) -> None:
```

I think my preference would be for the above three rules to be applied in every scope. I have no idea about the feasibility of this. Thank you for looking into it!

**Comment by bluetech** [comment]
> I'm currently looking into what it takes to make COM812 compatible with the formatter. It sounds like a simple change, but it's a very expensive token-based rule.

Curious, in my mental model token-based rules were supposed to be faster than AST based rules, but I guess I need to update :)

BTW, while flake8-commas (on which the ruff implementation is based) is token-based, there is also https://github.com/asottile/add-trailing-comma which IIRC is AST based, maybe it's a better base for this rule.

**Comment by bluetech** [comment]
> Can you explain how you add the parameter that it changes two lines? The formatting I'm thinking of either change one line or all parameter lines.

When not using a formatter (my case), would just another line. With a formatter (only with magic trailing comma?), I would prefer that format to never exist, should use the exploded (line-per-parameter) format instead.

**Comment by MichaReiser** [comment]
> Curious, in my mental model token-based rules were supposed to be faster than AST based rules, but I guess I need to update :)

It always depends. AST based rule have a high upfront cost because it requires parsing and building the AST. But that's something done once for all rules, so it then becomes "cheaper" for a single rule. 

Token based rules can be very fast and they tend to be faster than string based rules. However, they can be slow if they require a lot of context tracking. 


> When not using a formatter (my case), would just add another line. With a formatter (only with magic trailing comma?), I would prefer that format to never exist, should use the exploded (line-per-parameter) format instead.

Can you give me an example?

Like, would you write this as:

```python
def args(
    aaaaaaaa, bbbbbbbbb, cccccccccc, ddddddddd, eeeeeeee, ffffff, gggggggggggg, hhhh, 
	iiii
):
    pass
```

**Comment by bluetech** [comment]
> Like, would you write this as:

Yes, though of course with tab -> 4 spaces, and a trailing comma on `iiii`, that is

```py
def args(
    aaaaaaaa, bbbbbbbbb, cccccccccc, ddddddddd, eeeeeeee, ffffff, gggggggggggg, hhhh, 
    iiii,
):
    pass
```

though realistically when this happens I prefer to explode.

**Comment by bluetech** [comment]
(Just to be clear, I think the current situation where COM812 is not compatible with the formatter is *worse* than a situation where COM812 doesn't enforce comma in this case, so I think #10111 *is* an improvement. I'm just thinking if somehow we can have both).

**Comment by jagerber48** [comment]
Ok, so here area all the possible types of formats I could imagine someone wanting on their function signatures. The principles here apply to function calls and long literals as well.

```
# A) SINGLE LINE, SHARE WITH DEF
# Breaks line-too-long, Trailing comma unhelpful here
def args(aaaaaaaa, bbbbbbbbb, cccccccccc, ddddddddd, eeeeeeee, ffffff, gggggggggggg, hhhh, iiii):
    pass

# B) MANY PARAMS PER LINE, SHARE WITH DEF, 4-SPACE WRAP
# Trailing comma unhelpful here
def args(aaaaaaaa, bbbbbbbbb, cccccccccc, ddddddddd, eeeeeeee, ffffff, gggggggggggg, 
    hhhh, iiii):
    pass

# C) MANY PARAMS PER LINE, SHARE WITH DEF, MATCHING WRAP 
# Trailing comma unhelpful here
def args(aaaaaaaa, bbbbbbbbb, cccccccccc, ddddddddd, eeeeeeee, ffffff, gggggggggggg, 
         hhhh, iiii):
    pass

# D) SINGLE LINE, DON'T SHARE WITH DEF
# Breaks line-too-long, trailing comma unhelpful here
def args(
    aaaaaaaa, bbbbbbbbb, cccccccccc, ddddddddd, eeeeeeee, ffffff, gggggggggggg, hhhh, iiii
):
    pass

# F) MANY PARAMS PER LINE, DON'T SHARE WITH DEF, NO TRAILING COMMA
# loud vcs
def args(
    aaaaaaaa, bbbbbbbbb, cccccccccc, ddddddddd, eeeeeeee, ffffff, gggggggggggg, hhhh, 
    iiii
):
    pass

# G) MANY PARAMS PER LINE, DON'T SHARE WITH DEF, TRAILING COMMA
def args(
    aaaaaaaa, bbbbbbbbb, cccccccccc, ddddddddd, eeeeeeee, ffffff, gggggggggggg, hhhh, 
    iiii,
):
    pass

# H) SINGLE PARAM PER LNE, DON'T SHARE WITH DEF, NO TRAILING COMMA
# loud vcs
def args(
    aaaaaaaa, 
    bbbbbbbbb, 
    cccccccccc, 
    ddddddddd, 
    eeeeeeee, 
    ffffff, 
    gggggggggggg, 
    hhhh, 
    iiii
):
    pass

# I) # SINGLE PARAM PER LNE, DON'T SHARE WITH DEF, TRAILING COMMA
def args(
    aaaaaaaa, 
    bbbbbbbbb, 
    cccccccccc, 
    ddddddddd, 
    eeeeeeee, 
    ffffff, 
    gggggggggggg, 
    hhhh, 
    iiii,
):
    pass

# J) SINGLE PARAM PER LNE, SHARE WITH DEF
# Trailing comma unhelpful here
def args(aaaaaaaa, 
         bbbbbbbbb, 
         cccccccccc, 
         ddddddddd, 
         eeeeeeee, 
         ffffff, 
         gggggggggggg, 
         hhhh, 
         iiii):
    pass

```

Many of these "break" some rule amongst "don't allow line too long" and "don't cause two line-diffs when a trailing comma would make the diff only be one line". Also, convention seems to dictate that formats like people don't like formats like B), C) and J) (sharing a line with `def` even when multiple lines are present), so I'll forbid those also. If we want all those constraints to be followed then what remains are

```
# A) SINGLE LINE, SHARE WITH DEF
# Only valid for short enough line
def args(aaaaaaaa, bbbbbbbbb, cccccccccc, ddddddddd, eeeeeeee, ffffff, gggggggggggg, hhhh, iiii):
    pass
    
# D) SINGLE LINE, DON'T SHARE WITH DEF
# Only valid for short enough line
def args(
    aaaaaaaa, bbbbbbbbb, cccccccccc, ddddddddd, eeeeeeee, ffffff, gggggggggggg, hhhh, iiii
):
    pass

# G) MANY PARAMS PER LINE, DON'T SHARE WITH DEF, TRAILING COMMA
def args(
    aaaaaaaa, bbbbbbbbb, cccccccccc, ddddddddd, eeeeeeee, ffffff, gggggggggggg, hhhh, 
    iiii,
):
    pass

# I) # SINGLE PARAM PER LNE, DON'T SHARE WITH DEF, TRAILING COMMA
def args(
    aaaaaaaa, 
    bbbbbbbbb, 
    cccccccccc, 
    ddddddddd, 
    eeeeeeee, 
    ffffff, 
    gggggggggggg, 
    hhhh, 
    iiii,
):
    pass
```
I brought A and D back, only valid for short enough lines. I think D is valid so long as you are going to roll over to I instead of G if the line gets too long. If D is going to roll over to G when the line is too long then you should consider that a "single line case of G".

So I think it boils down to two options:

- Do you want multi-line signatures to wrap to G) or I)? 
- Do you want to use trailing commas in cases where they can reduce diff sizes in the event a line gets rolled over to a new line when a param/arg is added?

**Comment by jagerber48** [comment]
Yeah, to summarize, I think the simplest thing (and what people who use both `COM812` and the formatter get right now) is to use A) when the line isn't too long, but to immediately roll over to I) when the line gets too long.

But perhaps this is overly restrictive and there should be a way to choose some of the intermediate formats in addition?

**Comment by MichaReiser** [comment]
Changing the formatter isn't the solution in my view and changing the formatter's behavior based on the enabled lint rules would make it difficult to understand why the formatter behaves a certain way.

**Comment by MichaReiser** [comment]
Okay, I recommend that we leave this as is for now as documented on the [PR](https://github.com/astral-sh/ruff/pull/10111). 

* The formatter's formatting is widely supported in the community (proven by Black) and I don't want to make the formatter's behavior dependent on whether a certain lint rule is enabled or not
* We recommend disabling `COM812` if you use the formatter: The formatter takes care of the formatting for you
* You can use `COM812` and the formatter to catch the "all arguments on one line" formatting and add the trailing comma, but this is an unsupported feature, and the warning expresses that. 
* You can use `COM812` without the formatter
* I don't consider adding a new option as a solution because it doesn't solve the issue that the rule isn't compatible with the formatter. It simply makes the rule incompatible when using a certain set of options, leaving the situation unchanged.

**Comment by bluetech** [comment]
@MichaReiser really appreciate your detailed consideration of this issue.

Though I admit to being a bit confused at the outcome:

- The current situation is that COM812 and formatter are officially incompatible, possibly this even becomes a hard error in the future.

- This issue asks them to not be incompatible. If I understand the reasoning from @jagerber48 and @bittner (and myself) correctly, the reason why we want COM812 + formatter is to avoid the "SINGLE LINE, DON'T SHARE WITH DEF" case.

- The PR makes COM812 + formatter compatible, by making COM812 not complain about the "SINGLE LINE, DON'T SHARE WITH DEF" case.

It therefore seems to that the PR wouldn't actually satisfy what's being requested in this issue, and actually make the situation worse than the current situation for people *not* using the formatter (because they want a comma there).

Really sorry if I'm misunderstanding something and just wasting your time...

**Comment by bluetech** [comment]
Hmm @MichaReiser  just realized when you said "I recommend that we leave this as is for now" you meant the issue (i.e. close it) not the PR. So yeah I agree :)

**Comment by jagerber48** [comment]
@MichaReiser ok, I think the suggestion to disable `COM812` and just use the formatter will indeed give me outputs that are satisfactory. I think the expected behaviors are a bit confusing though and the documentation wasn't very clarifying for me.  Actually I generally don't see much documentation about what changes I should expect from the formatter? Maybe I should check documentation for `black`?

In any case, it seems like the formatter will:

(1) If the line is too long to have params/args share with def then it will
(2) try to put all parameters on a single line after `def(` and before `):` on a third line. It does this WITHOUT a trailing comma
(3) If the line is still too long then it breaks out each param/arg onto its own line WITH a trailing comma

Under my current workflow I will never get output formatted like (2) because of the lack of trailing comma. But if I were to disable the lint rule `COM812` I would sometimes get output like (2). That's ok with me, I don't have anything against (2).

So for me I would be happy to disable `COM812`. The reason I didn't want to do that was I thought it would allow output like
```
def args(
    aaaaaaaa, 
    bbbbbbbbb, 
    cccccccccc, 
    ddddddddd, 
    eeeeeeee, 
    ffffff, 
    gggggggggggg, 
    hhhh, 
    iiii
):
    pass
```
without a trailing comma. That is, since `COM812` was disabled I didn't expect anything to be there to enforce the presence of the trailing comma. But, it seems the formatter on its own already enforces the trailing comma in this case.

I would be helped by more clear documentation pointing out that running the formatter will already enforce trailing commas in the single param/arg per line format. Though the gotcha edge case here, as @bluetech highlights, is that the formatter will NOT enforce the trailing comma in the multiple params on one line not shared with `def` case. This is ok from a VCS perspective since if parameters are added the signature would be totally reformatted to the single param per line style. This is ok if you're totally relying on the formatter though because you'll never be outside one of the three possible cases.

@bluetech indicates that they want the trailing comma so that they can extend multi-params not shared with `def` onto more than one line. But in this case they are operating outside the rules of the formatter so they should not use the formatter. In this case I think `COM812` alone would suffice.

____

To summarize, I think status quo is ok for me, but better documentation would be helpful. Especially in light of the fact that the multi-parameters per line not shared with `def` is kind of a funny edge case in my view.

**Comment by bittner** [comment]
I believe the formatter does a sensible job.

To solve the contradiction, `COM812` should accept case "D" from above. Isn't that all there is to do?

(Or don't I get the point of all the numerous cases from above. Frankly, I've never seen the other cases caused by the formatter.)

I use both the linter and the formatter, because they serve different jobs. I have all linter rules enabled, I try to make exceptions rare and file-based. Ruff is great for this, giving actionable feedback collected from years of investment of a significant number of people in the Python community.

A note aside, I try to stay away from `--fix` and also review changes made by the formatter. It makes sense more often than expected to take your time and come up with "creative solutions" that are easier to read and more intuitive from a Clean Code point of view.

**Comment by bittner** [comment]
I can only reiterate that using both the formatter and the linter is not a strange idea. This issue/question is all about using both in a way that wouldn't contradict each other.

The multi-line multi-parameter per line case is unusual. Black would never accept it. The single-line multi-parameter case is generated by the formatter, which is a behavior we have seen by Black.

We should get the linter not complain about the generated formatting. With or without a linter rule setting. The current state is not satisfactory. And suggesting to stop using both the formatter and the linter together is not satisfactory either.

**Comment by charliermarsh** [comment]
Yeah, using the formatter and the linter together is great and very much intended. As you say, they solve different problems! I don't have a solution here, but adding some more color from my perspective... The issue in this case is that this rule pre-dated the formatter -- if the rule didn't exist today, we likely wouldn't add it to Ruff, since these kinds of linter-formatter conflicts create a lot of problems for users.

One answer would be to deprecate and remove the rule. But we've shied away from doing that, and instead we've:

- Documented the rules that conflict, and warned against them.
- Held back from adding _new_ rules that cause linter-formatter conflicts.
- In some cases, modified rules to make them non-conflicting.

Modifying the rule to allow Case (D) above would be a reasonable compromise, and would make it non-conflicting, but it'd _probably_ also make it redundant, and my guess is that users of this rule actually _like_ that it enforces that style.

**Comment by kozlek** [comment]
As explained in flake8-commas and ruff doc, COM812 was created to reduce diff size and it's working well.
It's approach is more aggressive than the current formatting style of ruff's formatter which allow to group all arguments on a single line before exploding all arguments line by line.

In real world use cases, you might start with a few number of arguments and then add more afterwards.
With COM812, you have two possible states:
- the compact with everything on a single line
- the diff friendly one, with everything on a separate line

Without COM812 and with ruff formatter, you get one more state in between:
- a less compact state with the function name on the first line, all the arguments on the second line and the return annotation on the third

Each state transition is painful during code review because you need to read again everything to figure out what changed so COM812 really helps by reducing the number of states.
I'm not trying to convince anyone to actually adopt this rule, but I simply emphasise why it's important to keep it untouched because there are users who rely on it, as @charliermarsh mentioned.

As @MichaReiser pointed, this opinion is not compatible with the current formatter style.
As a long time (convinced) user of COM812, this bothers me but I had the same problem with Black and the numerous benefits of ruff replacing black / isort / flake8 + having an uncompromising formatting style are worth it 🙂
For this reason, I'm going disable the rule on my projects.

Regarding what comes next, I have some questions:

- while I see the clear benefits of having an uncompromising formatter, would it be acceptable in the future to have an option in the formatter to remove the "the less compact" in mention earlier ? 
- as of today, ruff's formatting style seems deeply linked to Black style. Would it be possible in the future to suggest formatter style changes, similar to PEP ?

Thanks for all your explanations and for your time working on ruff, you keep saving us more and more time every day coding ❤️

**Comment by bittner** [comment]
I agree with @kozlek's point of the cognitive overhead when code is reformatted. "The less compact" variant has its value, though, just for the same reason. Let me explain:

- As an intermediate step, "the less compact" variant makes a less intrusive change than splitting all argument up into separate lines, immediately.
  - The number of lines doesn't change significantly with this change.
  - The change is limited to what's necessary at that moment.
- When "the less compact" variant gets too long on the line and the formatter starts to split up the arguments, then the leading and trailing parentheses are already in the right position.
  - The position of the parentheses doesn't change, which limits the transition to just the arguments.

The formatter's approach **minimizes the disruptive nature** of the changes.

> have an option in the formatter to remove the "the less compact" in mention earlier

In a certain sense, this is already possible today.

1. Simply run `ruff format` and `ruff lint` afterwards, with COM812 activated.
2. When the formatter generates "the less compact" variant, you'll get a linter complaint about a missing trailing comma.
3. Add the comma and run the `ruff format` again; it will reformat the code using the "diff friendly" variant.

If I'm not totally wrong this two-stage formatting is exactly the behavior of Black, too.

My take is, the formatter does a decent job, also for people who really don't want to compromise with COM812. If we had a **configuration switch** for the _linter_ that would allow a **less strict interpretation** of COM812 (for accepting "the less compact" variant) wouldn't we fix today's situation, which would allow us to remove the conflict warning?

**Comment by MichaReiser** [comment]
> My take is, the formatter does a decent job, also for people who really don't want to compromise with COM812. If we had a configuration switch for the linter that would allow a less strict interpretation of COM812 (for accepting "the less compact" variant) wouldn't we fix today's situation, which would allow us to remove the conflict warning?

It would solve the conflict, but I don't see the difference to disabling the lint and only use the formatter (except that the warning doesn't show up by default when using `--select ALL`). 

@AlexWaygood's currently figuring out how we want to [categorize Ruff's rules](#1774) in the future. This work also touches on if Ruff should support rules that are incompatible with the formatter (probably?) and, if so, how the user experience looks like. 

That's why I want to hold off from making any changes to `COM812`.

**Comment by ZetaDaemon** [comment]
Ran into this when using the `skip-magic-trailing-comma` rule where the formatter would remove the trailing comma while the linter complains about the missing comma
https://play.ruff.rs/407b6f2a-9e84-4e8b-a4d1-eca87138211d is an example of how I ran into it

**Comment by GutiCW** [comment]
Thanks for bringing this up - I'll just leave my opinion here: ;-)

It just seems inconsistent formatting to different styles:
![Image](https://github.com/user-attachments/assets/0764767e-ff3e-4e69-9c89-dc47a2380879)
[Ruff playground](https://play.ruff.rs/f4651b3d-a0dc-4f6e-9054-3c220b698039)

To me a minimal diff is most import, therefore (and as already mentioned earlier) from [missing-trailing-comma](https://docs.astral.sh/ruff/rules/missing-trailing-comma/):
>The presence of a trailing comma can reduce diff size when parameters or elements are added or removed from function calls, function definitions, literals, etc.

**Comment by ansipunk** [comment]
I don't care. I want formatter, linter and COM812 live together. And I want this:
```
def func(
    foo, bar, baz
):
```
formatted into this:
```
def func(
    foo, bar, baz,
):
```
I understand that there are programming languages that don't support the trailing comma hence it's not that common to have it when the arguments on one line, however, it looks incomplete and feels inconsistent with, let's say,
```
def func(
    foo,
    bar,
    baz,
):
```
I understand that every programmer has their own style preferences that don't necessarily adhere with mine and none of them is really the "correct" way to do stuff, however, I'd like the formatter to do stuff the way I like. If black style is not necessarily compatible with popular (like flake8-comma or some popular pylint defaults), then it's wrong and it has to be configurable.

**Comment by PeterJCLaw** [comment]
> An option that I would be supportive of is to add a config option to COM812 or introduce a new rule that's compatible with the formatter. This way, users that just want to deny
> 
> ```python
> def func(
>     foo, bar, baz
> ):
> ```
> 
> can use it. Everyone else can keep using the rule as is, but they see the compatibility warning.

This feels reasonable to me. Another option here might be to break COM812 (and flake8-commas' C812) into separate errors by context -- so have one for calls, one for function defs, one for class defs, one for collection literals, etc. This would avoid introducing a new setting while also providing users with more granularity.

I'm personally in the camp of wanting to enable both C812/COM812 and the formatter, so I'd love to see an option which let me do that _and then silence the warning_. Even better if there's a way to not need to run the fixer & the formatter again, but I'm less worried about that (as it's easily automated).
Something I've considered for that is whether including `"COM812"` explicitly in the list of selected lints (rather than just a result of its prefix matching) should at least silence the warning? (If I've missed an existing way to silence the warning I'm all ears!)

**Comment by akx** [comment]
It would be nice to be able to silence the warning, at least.

There's already `unfixable`, but... `unwarnable`? 😅 That said, just `unwarnable = ["COM812"]` would not quite be enough, because there could be multiple distinct warnings about a single rule. (Should warnings emitted by Ruff have unique IDs, like `missing-trailing-comma-unstable-with-formatter`?)

**Comment by MichaReiser** [comment]
We should just implement https://github.com/astral-sh/ruff/issues/9216#issuecomment-2904749624 or split the rule into two instead of implementing any complex workarounds

**Comment by XuehaiPan** [comment]
I've been hitting this much harder since AI coding agents (Claude Code, Codex, Cursor, etc.) started generating most of the code in my repos. They frequently produce over-wrapped expressions with unnecessary trailing commas, and the formatter can't normalize them to a deterministic output in one pass:

```python
# AI-generated (over-wrapped with unnecessary trailing comma)
result = some_function(
    arg1,
    arg2,
)

# Expected (collapsed, no trailing comma needed)
result = some_function(arg1, arg2)
```

My setup:

```toml
[tool.ruff.format]
skip-magic-trailing-comma = false

[tool.ruff.lint]
select = ["COM", "..."]  # flake8-commas + 20 other rule groups
```

```yaml
# .pre-commit-config.yaml
- repo: https://github.com/astral-sh/ruff-pre-commit
  rev: v0.15.10
  hooks:
    - id: ruff-check
      args: [--fix, --exit-non-zero-on-fix]
    - id: ruff-format
      args: [--exit-non-zero-on-format]
```

With `skip-magic-trailing-comma = false` and `COM812` enabled, `ruff format` sees the trailing comma and refuses to collapse. The only cleanup path is a multi-pass workaround:

```bash
# One-shot collapse
ruff format --config="format.skip-magic-trailing-comma=true" .

# Then converge COM812 + formatter — takes up to 3 passes
for i in {1..3}; do pre-commit run --all-files && break; done
```

I tried to automate this inside pre-commit but it turned out to be impossible:

- **Separate hooks oscillate**: the collapse pass removes trailing commas, `COM812` adds them back, `ruff format` re-expands — same fixes and reformats every pass, never converges.
- **Setting `skip-magic-trailing-comma = true` in `pyproject.toml`** also oscillates for the same reason. Ruff already warns about this.
- **A single combined system hook** that runs the full pipeline internally does converge, but writes intermediate states to disk multiple times per file. When running `pre-commit run --all-files`, it touches every Python file in the repo even though the final state is identical to the original — causing stale reads in AI agents and unnecessary I/O churn:

    ```yaml
    # .pre-commit-config.yaml
    #   - repo: https://github.com/astral-sh/ruff-pre-commit
    #     rev: v0.15.10
    #     hooks:
    #       - id: ruff-check
    #         args: [--fix, --exit-non-zero-on-fix]
    #       - id: ruff-format
    #         args: [--exit-non-zero-on-format]
    - repo: local
      hooks:
        - id: ruff
          name: ruff
          language: system
          entry: bash -c '
            python3 -m ruff format --config="format.skip-magic-trailing-comma=true" "$@" 2>/dev/null;
            for i in 1 2 3; do
              python3 -m ruff check --fix "$@" 2>/dev/null;
              python3 -m ruff format "$@" 2>/dev/null;
            done;
            python3 -m ruff check "$@" 2>/dev/null && python3 -m ruff format --check "$@" 2>/dev/null
            ' --
          types: [python]
    ```

Pre-commit expects each hook to be idempotent in one pass, but the current formatter + `COM812` combination fundamentally requires multiple passes to reach a stable state.

It would be nice to have `ruff format` remove unnecessary trailing commas and produce deterministic output in a single invocation — idempotent, no intermediate disk writes, no multi-pass convergence loop.

---

Reproducible example — all three cases produce different output depending on the input trailing comma, rather than converging to one deterministic form:

```python
# Case 1: over-wrapped with trailing comma (AI-generated)
# `ruff format` keeps it multi-line because of the magic trailing comma
result = some_function(
    arg1,
    arg2,
)

# Case 2: over-wrapped without trailing comma
# `ruff format` collapses it to Case 3 — good
result = some_function(
    arg1,
    arg2
)

# Case 3: single line without trailing comma — the desired output
result = some_function(arg1, arg2)
```

Case 2 correctly collapses to Case 3, but Case 1 stays multi-line because the trailing comma prevents collapsing. The formatter produces different output for the same expression depending on whether a trailing comma is present.

```python
# Case 4: genuinely long — must stay multi-line and keep the trailing comma
result = some_function(
    first_argument,
    second_argument,
    third_argument,
    fourth_argument,
    fifth_argument,
)
```

```python
# Case 5: doesn't fit on one line, formatter produces compact multi-line
# without trailing comma — this is the existing conflict from this thread
result = some_function(
    first_argument, second_argument, third_argument, fourth_argument
)

# Expected: COM812 adds trailing comma to reduce git diff when args change,
# then formatter re-expands to one-arg-per-line
result = some_function(
    first_argument,
    second_argument,
    third_argument,
    fourth_argument,
)
```

Cases 4 and 5 are genuinely multi-line — the trailing comma is useful there to reduce diff size. The problem is Case 1 where an unnecessary trailing comma prevents collapsing to Case 3.

**Comment by XuehaiPan** [comment]
I opened a PR to resolve the issue I posted in https://github.com/astral-sh/ruff/issues/9216#issuecomment-4229572901.

- #24572

**Comment by WhyNotHugo** [comment]
Mirroring my comment from https://github.com/astral-sh/ruff/pull/24572 for better visibility, I'd love to see a way to have Ruff always properly auto-format code, without needing to sprinkle "magic" commas to trigger reformatting.

In particular, the following is a weird mix which is always confusing at a glance:

```
result = some_function(
    another(first_argument, second_argument, third_argument), fourth_argument
)
```

I'd much rather see

```
result = some_function(
    another(first_argument, second_argument, third_argument),
    fourth_argument,
)
```

I.e.: I'd want all parameters to either be on the same line as the function name, or one-per-line. The current approach yields a half-formatted mix which is the worst of all worlds: any change has a large diff and at a glance it looks like there's a single parameter being passed.

**Comment by MichaReiser** [comment]
@WhyNotHugo you can track https://github.com/astral-sh/ruff/issues/24611 which requests this specific style change.
